google/cloud/bigtable/iterators.py

```diff
@@ -0,0 +1,135 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import annotations
+
+from typing import (
+    cast,
+    AsyncIterable,
+)
+import asyncio
+import time
+import sys
+
+from google.cloud.bigtable._row_merger import _RowMerger
+from google.cloud.bigtable_v2.types import RequestStats
+from google.api_core import exceptions as core_exceptions
+from google.cloud.bigtable.exceptions import RetryExceptionGroup
+from google.cloud.bigtable.exceptions import IdleTimeout
+from google.cloud.bigtable.row import Row
+
+
+class ReadRowsIterator(AsyncIterable[Row]):
+    """
+    Async iterator for ReadRows responses.
+    """
+
+    def __init__(self, merger: _RowMerger):
+        self._merger_or_error: _RowMerger | Exception = merger
+        self.request_stats: RequestStats | None = None
+        self.last_interaction_time = time.time()
+        self._idle_timeout_task: asyncio.Task[None] | None = None
+
+    async def _start_idle_timer(self, idle_timeout: float):
+        """
+        Start a coroutine that will cancel a stream if no interaction
+        with the iterator occurs for the specified number of seconds.
+
+        Subsequent access to the iterator will raise an IdleTimeout exception.
+
+        Args:
+          - idle_timeout: number of seconds of inactivity before cancelling the stream
+        """
+        self.last_interaction_time = time.time()
+        if self._idle_timeout_task is not None:
+            self._idle_timeout_task.cancel()
+        self._idle_timeout_task = asyncio.create_task(
+            self._idle_timeout_coroutine(idle_timeout)
+        )
+        if sys.version_info >= (3, 8):
+            self._idle_timeout_task.name = "ReadRowsIterator._idle_timeout"
+
+    def active(self):
+        """
+        Returns True if the iterator is still active and has not been closed
+        """
+        return isinstance(self._merger_or_error, _RowMerger)
+
+    async def _idle_timeout_coroutine(self, idle_timeout: float):
+        """
+        Coroutine that will cancel a stream if no interaction with the iterator
+        in the last `idle_timeout` seconds.
+        """
+        while self.active():
+            next_timeout = self.last_interaction_time + idle_timeout
+            await asyncio.sleep(next_timeout - time.time())
+            if (
+                self.last_interaction_time + idle_timeout < time.time()
+                and self.active()
+            ):
+                # idle timeout has expired
+                await self._finish_with_error(IdleTimeout("idle timeout expired"))
```

Comment:
Maybe this can be more descriptive, maybe something like: Timed out waiting for the next Row to be consumed.idle timeout is <idle_timeout number>